File: Scala/algorithmic_toolbox/src/main/scala-2.12/divideandconquer/CountInversions.scala
package divideandconquer

import scala.annotation.tailrec

object CountInversions {
  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def mergeSortedArrays(first: List[Int],
                                second: List[Int],
                                length1: Int,
                                length2: Int,
                                inversions: Long): (List[Int], Long) = {
    @tailrec
    def merge(xs: List[Int],
              xsLength: Int,
              ys: List[Int],
              mergedList: List[Int],
              accInversions: Long): (List[Int], Long) = (xs, ys) match {
      case (Nil, Nil) => (mergedList.reverse, accInversions)
      case (Nil, y :: yss) => merge(Nil, xsLength, yss, y :: mergedList, accInversions)
      case (x :: xss, Nil) => merge(xss, xsLength - 1, Nil, x :: mergedList, accInversions)
      case (x :: xss, y :: yss) =>
        if (x <= y) merge(xss, xsLength - 1, ys, x :: mergedList, accInversions)
        else merge(xs, xsLength, yss, y :: mergedList, accInversions + xsLength)
    }
    merge(first, length1, second, Nil, inversions)
  }

  def countInversions(array: List[Int], length: Int): (List[Int], Long) = {
    if (length <= 1) (array, 0L)
    else {
      val middle: Int = length / 2
      val (first, second): (List[Int], List[Int]) = array.splitAt(middle)
      val (length1, length2): (Int, Int) = (middle, length - middle)
      val (sortedFirst, inversionsInFirst): (List[Int], Long) = countInversions(first, length1)
      val (sortedSecond, inversionsInSecond): (List[Int], Long) = countInversions(second, length2)
      mergeSortedArrays(sortedFirst, sortedSecond, length1, length2, inversionsInFirst + inversionsInSecond)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val lst: List[Int] = convertToIntList(reader.next())
    val (_, result): (List[Int], Long) = countInversions(lst, n)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/divideandconquer/count_inversions.ml
open Batteries

let convert_to_intlist line = List.map int_of_string (Str.split (Str.regexp " ") line)


let merge_sorted_arrays first second length1 inversions =
    let rec merge xs xs_length ys acc total_inversions = match (xs, ys) with
        | [], [] -> List.rev acc, total_inversions
        | [], y :: yss -> merge [] xs_length yss (y :: acc) total_inversions
        | x :: xss, [] -> merge xss (xs_length - 1) [] (x :: acc) total_inversions
        | x :: xss, y :: yss ->
            if (x <= y) then merge xss (xs_length - 1) ys (x :: acc) total_inversions
            else merge xs xs_length yss (y :: acc) (total_inversions + xs_length)
    in merge first length1 second [] inversions


let rec count_inversions lst length =
    if length <= 1 then lst, 0
    else
        let middle = length / 2 in
        let first, second = List.split_at middle lst in
        let length1, length2 = middle, length - middle in
        let sorted_first, inversions_first = count_inversions first length1 in
        let sorted_second, inversions_second = count_inversions second length2 in
        merge_sorted_arrays sorted_first sorted_second length1 (inversions_first + inversions_second)


let () =
    let n = read_int() in
    let lst = convert_to_intlist (read_line()) in
    let _, result = count_inversions lst n in
    print_int result;
    print_newline()
